contracts/cw721-stake/src/contract.rs

```diff
@@ -0,0 +1,1821 @@
+            });
+        }
+
+        let sender = deps.api.addr_validate(&wrapper.sender)?;
+        REWARD_BALANCE.update(deps.storage, |old| {
+            old.checked_add(wrapper.amount).map_err(StdError::overflow)
+        })?;
+
+        Ok(Response::new()
+            .add_attribute("action", "fund")
+            .add_attribute("from", &sender)
+            .add_attribute("amount", wrapper.amount))
+    } else {
+        Err(ContractError::NothingToFund {})
+    }
+}
+
+pub fn execute_stake(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    wrapper: Cw721ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if info.sender != config.nft_address {
+        return Err(ContractError::InvalidToken {
+            received: info.sender,
+            expected: config.nft_address,
+        });
+    }
+
+    let sender = deps.api.addr_validate(&wrapper.sender)?;
+    STAKED_NFTS_PER_OWNER.update(
+        deps.storage,
+        &sender,
+        env.block.height,
+        |nft_collection| -> StdResult<HashSet<String>> {
+            let mut updated_nft_collection = nft_collection.unwrap_or_default();
+            updated_nft_collection.insert(wrapper.token_id.clone());
+            Ok(updated_nft_collection)
+        },
+    )?;
+
+    TOTAL_STAKED_NFTS.update(
+        deps.storage,
+        env.block.height,
+        |total_staked_nfts| -> StdResult<HashSet<String>> {
+            let mut updated_total_staked_nfts = total_staked_nfts.unwrap_or_default();
+            updated_total_staked_nfts.insert(wrapper.token_id.clone());
+            Ok(updated_total_staked_nfts)
+        },
+    )?;
+
+    let hook_msgs = stake_hook_msgs(deps.storage, sender.clone(), wrapper.token_id.clone())?;
+    Ok(Response::new()
+        .add_submessages(hook_msgs)
+        .add_attribute("action", "stake")
+        .add_attribute("from", sender)
+        .add_attribute("token_id", wrapper.token_id))
+}
+
+pub fn execute_unstake(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    token_id: String,
+    reward_wallet_address: Option<String>,
+) -> Result<Response, ContractError> {
+    let validated_reward_wallet = reward_wallet_address
+        .map(|h| deps.api.addr_validate(&h))
+        .transpose()?;
+    let sender = info.sender.clone();
+    let config = CONFIG.load(deps.storage)?;
+    let sender_staked_nft_collection = STAKED_NFTS_PER_OWNER
+        .may_load(deps.storage, &sender)?
+        .unwrap_or_default();
+    if !sender_staked_nft_collection.contains(&token_id) {
+        return Err(ContractError::Unauthorized {});
+    }
+    let balance = REWARD_BALANCE.load(deps.storage).unwrap_or_default();
+    let total_staked_nfts = TOTAL_STAKED_NFTS.load(deps.storage)?;
+    let amount_to_claim = balance
+        .checked_div(Uint128::from(
+            u128::try_from(total_staked_nfts.len()).unwrap(),
+        ))
+        .map_err(StdError::divide_by_zero)?;
```

Comment:
With these calculations I can drain the rewards pool.

Say there are 100 reward tokens and 10 NFTs. I have one staked.

1. I unstake my NFT and receive 10 rewards tokens => 90 rewards tokens, 9 NFTs.
2. I stake my NFT => 90 rewards tokens, 10 NFTs.
3. I unstake my NFT and receive 9 rewards tokens => 81 rewards tokens, 9 NFTs.
4. ... repeat

The way that the cw20-stake contract gets around this is that when you stake your balance increases by the proportional ownership and not the amount of tokens you have staked. From `execute_stake`:

```rust
    let amount_to_stake = if staked_total == Uint128::zero() || balance == Uint128::zero() {
        amount
    } else {
        staked_total
            .checked_mul(amount)
            .map_err(StdError::overflow)?
            .checked_div(balance)
            .map_err(StdError::divide_by_zero)?
    };
```

I _think_ this is possible to do with NFT staking if we store `STAKED_BALANCE` and also NFT information and then use NFT information only for sending the correct NFT back and not for voting power calculations. This is tricky stuff.

Your choice on how to move forward here. I would be fine just removing rewards from this contract entirely though. We can start simple and then add them in later.

If we do choose to go the remove rewards route though it would be great to keep the `StakedValue` query around. The frontend uses this query for determining voting power and it would make our lives a little simpler over there. In this case it would just return the number of NFTs staked.

Reply:
After writing this review I realized my flash loan program had the same issue aha https://github.com/ezekiiel/cw-flash-loan/commit/3b77e6bc2c1c02f359c3430329c77917e3b9b3fc. You are not alone in running into this sort of thing.